main/src/com/google/refine/commands/history/ApplyOperationsCommand.java

```diff
@@ -59,26 +59,31 @@ public void doPost(HttpServletRequest request, HttpServletResponse response)
             respondCSRFError(response);
             return;
         }
-
-        Project project = getProject(request);
-        String jsonString = request.getParameter("operations");
         try {
-            ArrayNode a = ParsingUtilities.evaluateJsonStringToArrayNode(jsonString);
-            int count = a.size();
-            for (int i = 0; i < count; i++) {
-                if (a.get(i) instanceof ObjectNode) {
-                    ObjectNode obj = (ObjectNode) a.get(i);
+            Project project = getProject(request);
+            String jsonString = request.getParameter("operations");
 
-                    reconstructOperation(project, obj);
+            try {
+                ArrayNode a = ParsingUtilities.evaluateJsonStringToArrayNode(jsonString);
+                int count = a.size();
+                for (int i = 0; i < count; i++) {
+                    if (a.get(i) instanceof ObjectNode) {
+                        ObjectNode obj = (ObjectNode) a.get(i);
+
+                        reconstructOperation(project, obj);
+                    }
                 }
+            } catch (IllegalArgumentException e) {
+                respondCodeError(response, e.getMessage(), HttpServletResponse.SC_BAD_REQUEST, null);
+                return;
             }
 
             if (project.processManager.hasPending()) {
                 respond(response, "{ \"code\" : \"pending\" }");
             } else {
                 respond(response, "{ \"code\" : \"ok\" }");
             }
-        } catch (IOException e) {
+        } catch (Exception e) {
```

Comment:
Why was this catch widened? I think we should try to avoid catching `Exception` except in extreme last-chance cases

Reply:
Because `reconstructOperation` doesn't throw `IOException` when an invalid structure is encountered, but `IllegalArgumentException`. Catching all exceptions at the root of the command is the only way I can think of to make sure the backend responds with an appropriate JSON response to the frontend, as it isn't handled by the framework for us. Do you have examples of exceptions you would let through?